prov/efa/src/rxr/rxr_read.c

```diff
@@ -82,7 +82,47 @@ int rxr_locate_rma_iov_pos(struct fi_rma_iov *rma_iov, int rma_iov_count, size_t
 ssize_t rxr_read_mr_reg(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 {
 	size_t i;
-	int err;
+	int err, pkt_offset;
+	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_pkt_entry *pkt_entry_copy;
+
+	if (read_entry->x_entry_type == RXR_PKT_ENTRY) {
+		/* In this case, we are reading data in a pkt_entry.
+		 */
+		assert(read_entry->rma_iov_count == 1);
+
+		pkt_entry = read_entry->x_entry;
+		if (pkt_entry->mr) {
+			assert(read_entry->rma_iov[0].key == fi_mr_key(pkt_entry->mr));
+			return 0;
+		}
+
+		/* ooo and unexp packet entry's memory is not registered with device */
+		assert(pkt_entry->type == RXR_PKT_ENTRY_OOO ||
+		       pkt_entry->type == RXR_PKT_ENTRY_UNEXP);
+
+		pkt_offset = (char *)read_entry->rma_iov[0].addr - (char *)pkt_entry->pkt;
+		assert(pkt_offset > sizeof(struct rxr_base_hdr));
+
+		pkt_entry_copy = rxr_pkt_entry_clone(ep, ep->rx_readcopy_pkt_pool,
+						     pkt_entry, RXR_PKT_ENTRY_READ_COPY);
+		if (!pkt_entry_copy) {
+			FI_WARN(&rxr_prov, FI_LOG_CQ,
+				"readcopy pkt pool exhausted! Set FI_EFA_READCOPY_POOL_SIZE to a higher value!");
+			return -FI_ENOBUFS;
```

Comment:
I thought we were going to queue in this case? This will write an error (and crash if no EQ).

Reply:
Yes, changed to -FI_EAGAIN in new revision.